File: Cloud Sheets Samples/Google Sheets/javascript/googlesheets.js
/*
 * Cloud Sheets 2024 JavaScript Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Sheets in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudsheets
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */
 
const readline = require("readline");
const cloudsheets = require("@nsoftware/cloudsheets");

if(!cloudsheets) {
  console.error("Cannot find cloudsheets.");
  process.exit(1);
}
let rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});


class GoogleSheetsDemo {
    constructor() {
        this.oAuthScope = "https://www.googleapis.com/auth/spreadsheets https://www.googleapis.com/auth/drive";
        this.oAuthAuthUrl = "https://accounts.google.com/o/oauth2/auth";
        this.oAuthTokenUrl = "https://accounts.google.com/o/oauth2/token";
        this.gsheets = null;
        this.currentSpreadsheetId = "";
        this.selectedSpreadsheetName="";
        this.currentSheetId = -1;
        this.changesMade = false;
        this.currentSheet = null;
        this.rl = rl;

        this.running = true;
    }

    askQuestion(query) {
        return new Promise(resolve => {
            this.rl.question(query, resolve);
        });
    }
    reset = async () => 
    {
        this.currentSpreadsheetId="";
        await this.gsheets.listSpreadsheets();
    }
    getNumberInput= async (prompt) => {
        const input = (await this.askQuestion(prompt)).trim();
        const num = parseInt(input, 10);
        return (isNaN(num) || num < 0) ? null : num;
    }
    
    getConfirmation = async(question) =>
    {
        const answer = await this.askQuestion(question);
        return answer.toLowerCase() === 'y';
    }
    getCommand = async() => 
    {
        const prompt = this.currentSpreadsheetId 
        ? `${this.selectedSpreadsheetName}>`
        : 'GoogleSheets> ';
        return await this.askQuestion(prompt);
    }
    
    selectSpreadsheet = async(spreadsheetName) => 
    {
        try {
            await this.loadSpreadsheet(spreadsheetName);
            this.displaySpreadsheetMenu();
            
        }catch(error)
        {
            console.error("Error selecting spreadsheet:", error);
        }
    }
    loadSpreadsheet = async(spreadsheetName) =>
    {
        if(!spreadsheetName)
        {
            spreadsheetName = await this.askQuestion("Enter the Spreadsheet Name: ");
        }
        this.currentSpreadsheetId = await this.getSpreadsheetId(spreadsheetName);
        if(this.currentSpreadsheetId != null)
        {
            await this.gsheets.loadSpreadsheet(this.currentSpreadsheetId);
            this.selectedSpreadsheetName = spreadsheetName
        }
        else
        {
            await this.loadSpreadsheet()
        }
    }
    displaySpreadsheetMenu() {
        if (!this.selectedSpreadsheetName ==="") {
            console.log("No spreadsheet selected.");
            return;
        }
        console.log(
            "\nls\t-\tList Sheets\n" +
            "add\t-\tAdd Sheet\n" +
            "update\t-\tUpdate Sheet\n" +
            "delete\t-\tDelete Sheet\n" +
            "exit\t-\tReturn to Main Menu\n"
        );
    }
    displayMainMenu() {
        console.log(
                "?\t-\tHelp\n" +
                "ls\t-\tList Spreadsheets\n" +
                "sel\t-\tSelect Spreadsheet\n" +
                "cs\t-\tCreate Spreadsheet\n" +
                "us\t-\tUpdate Spreadsheet\n" +
                "del\t-\tDelete Spreadsheet\n" +
                "exp\t-\tExport Spreadsheet\n" +
                "q\t-\tQuit\n"
        );
    }
    listSheets = async() => 
    {
        const sheets = this.gsheets.getSheets()
        for (let i = 0; i < sheets.size(); i++) {
            console.log("\nId: " + sheets.get(i).getSheetId());
            console.log("Title: " + sheets.get(i).getTitle());
            console.log("Index: " + sheets.get(i).getSheetIndex());
            console.log("Total Columns: " + sheets.get(i).getTotalColumnCount());
            console.log("Total Rows: " + sheets.get(i).getTotalRowCount());
        }
    }

    getSpreadsheetId = async (spreadsheetName) =>
    {
        const spreadsheets = this.gsheets.getSpreadsheets();
        for (let i = 0; i < spreadsheets.size(); i++) {
            const name = spreadsheets.get(i).getName();
            if(name === spreadsheetName)
            {
                return spreadsheets.get(i).getId();
            }
        }
        if (!this.currentSpreadsheetId) {
            console.log(`No spreadsheet found with name: ${spreadsheetName}`);
            return null;
        }
    };
    listSpreadsheets = async () => {
        if(this.currentSpreadsheetId != "")
        {
            await this.listSheets();
            return;
        }
        const spreadsheets = this.gsheets.getSpreadsheets();
        const maxNameLength = 24;

        console.log("Name".padEnd(maxNameLength));
        console.log("-".repeat(20)); 

        for (let i = 0; i < spreadsheets.size(); i++) {
            const name = spreadsheets.get(i).getName();
            console.log(`${name}`);
        }
    };
    createSpreadsheet = async(spreadSheetName) => 
    {
        if(!spreadSheetName)
        {
            spreadSheetName = await this.askQuestion("Enter Spreadsheet name: ");
        }
        const locale = await this.askQuestion("Enter Spreadsheet locale (optional): ");
        const timezone = await this.askQuestion("Enter Spreadsheet timezone (optional): ");
        await this.gsheets.createSpreadsheet(spreadSheetName,locale,timezone);
        console.log(`Spreadsheet ${spreadSheetName} created successfully.`)
    };
    addSheet = async(sheetTitle) => 
    {
        if(this.selectedSpreadsheetName === "")
        {
            console.log("A spreadsheet must first be selected")
            return;
        }
        if(!sheetTitle)
        {
            sheetTitle = await this.askQuestion("Enter Sheet title: ");
        }
        await this.gsheets.addSheet(this.gsheets.getSheets().size()+1,sheetTitle);
        console.log("Sheet added successfully!")
    };

    updateSpreadsheet = async(spreadSheetName) => 
    {
        await this.loadSpreadsheet(spreadSheetName);
        const spreadsheets = this.gsheets.getSpreadsheets();
        for (let i = 0; i < spreadsheets.size(); i++) {
            if(this.currentSpreadsheetId === spreadsheets.get(i).getId())
            {
                const currentSpreadsheet = spreadsheets.get(i);
                const originalName = currentSpreadsheet.getName();
                const originalLocale=currentSpreadsheet.getLocale();
                const originalTimeZone=currentSpreadsheet.getTimeZone();

                const newName = await this.askQuestion(`Name [${originalName}]: `)
                const newLocale = await this.askQuestion(`Locale [${originalLocale}]: `)
                const newTimeZone = await this.askQuestion(`TimeZone [${originalTimeZone}]: `)

                this.gsheets.getSpreadsheets().get(i).setName(newName || originalName)
                this.gsheets.getSpreadsheets().get(i).setLocale(newLocale || originalLocale)
                this.gsheets.getSpreadsheets().get(i).setTimeZone(newTimeZone || originalTimeZone)

                await this.gsheets.updateSpreadsheet();
                console.log("Spreadsheet updated successfully!")
            } 
        }
        this.reset();
    };

    updateSheet = async() => 
    {
        if(this.selectedSpreadsheetName === "")
        {
            console.log("A spreadsheet must first be selected")
            return;
        }

        this.currentSheetId = await this.getNumberInput("Select a SheetID to update: ")
        await this.gsheets.loadSheet(this.currentSheetId);
        if(await this.getConfirmation("Change sheet name (Y/N)? "))
        {
            const newName = await this.askQuestion("Enter new sheet name: ")
            this.gsheets.getSheets().get(0).setTitle(newName);
            await this.gsheets.updateSheet();
        }
    };

    deleteSheet = async() => 
    {
        if(this.selectedSpreadsheetName === "")
        {
            console.log("A spreadsheet must first be selected")
            return;
        }
        this.currentSheetId = await this.getNumberInput("Select a SheetID to delete: ")
        await this.gsheets.loadSheet(this.currentSheetId);
        if(await this.getConfirmation("Delete Sheet (Y/N)? "))
        {
            await this.gsheets.deleteSheet(this.currentSheetId);
        }
    }
    deleteSpreadsheet = async(spreadSheetName) => 
    {   
        await this.loadSpreadsheet(spreadSheetName);
        await this.gsheets.deleteSpreadsheet(this.currentSpreadsheetId)
        await this.reset()
    }
    export = async() => 
    {
        await this.loadSpreadsheet();
        console.log("Supported file extensions: csv, ods, pdf, tsv, xlsx, zip");
        const fileSavePath = await this.askQuestion("Enter export location ex. (C:/Downloads/myFile.pdf) ")
        await this.gsheets.export(this.currentSpreadsheetId,fileSavePath)
        await this.reset()
    }
    exit = async()=>
    {
        this.currentSpreadsheetId ="";
        this.currentSheetId="";
        await this.gsheets.listSpreadsheets();
    }

    async main(args) {
        if (args.length < 4) {
            console.log("usage: node googlesheets.js -c client -s secret\n" +
                " -c The Id of the client assigned when registering the application.\n" +
                " -s The secret value for the client assigned when registering the application.");
            return;
        }

        try {
            const myArgs = this.parseArgs(args);
            this.gsheets = new CloudSheets.googlesheets();

            await this.authenticate(myArgs.get("c"), myArgs.get("s"));
            console.log("Loading Spreadsheets...");
            await this.gsheets.listSpreadsheets();
            this.currentSpreadsheetId ? this.displaySpreadsheetMenu() : this.displayMainMenu();

            while (this.running) {
                try {

                    const input= await this.getCommand();
                    const [command, ...args] = input.trim().split(/\s+/);

                    // Process commands
                    switch (command.toLowerCase().trim()) {
                        case 'ls':
                            await this.listSpreadsheets();
                            break;

                        case '?':
                            this.currentSpreadsheetId ? this.displaySpreadsheetMenu() : this.displayMainMenu();
                            break;
                        case 'sel':
                            await this.selectSpreadsheet(args[0]);
                            break;
                        case 'lssh':
                            await this.listSheets();
                            break;
                        case 'cs':
                            await this.createSpreadsheet(args[0]);
                            break;
                        case 'add': 
                            await this.addSheet(args[0]);
                            break;
                        case 'us': 
                            await this.updateSpreadsheet(args[0]);
                            break;
                        case 'update': 
                            await this.updateSheet();
                            break;
                        case 'del': 
                            await this.deleteSpreadsheet(args[0]);
                            break;
                        case 'delete': 
                            await this.deleteSheet();
                            break;
                        case 'exp': 
                            await this.export();
                            break;
                        case 'exit': 
                            await this.exit();
                            break;
                        case 'q':
                            console.log('Goodbye!');
                            this.rl.close();
                            this.running = false;
                            return;
                        default:
                            console.log('Invalid command. Please try again.');
                    }
                } catch (error) {
                    console.error('An error occurred:', error.message);
                    break;
                }
            }

        } catch (ex) {
            console.log(ex.message);
        }

    }

    async authenticate(client, secret) {
        if (client && secret) {
            const oauth = new CloudSheets.oauth();
            oauth.setClientId(client)
            oauth.setClientSecret(secret);
            oauth.setAuthorizationScope(this.oAuthScope);
            oauth.setServerAuthURL(this.oAuthAuthUrl);
            oauth.setServerTokenURL(this.oAuthTokenUrl);

            try {
                let authorization = await oauth.getAuthorization()
                await this.gsheets.setAuthorization(authorization);
            } catch (ex) {
                console.log(ex.message);
            }
        } else {
            console.log("Client Id and Secret cannot be null!");
        }
    }
    parseArgs(args) {
        const myArgs = new Map();
        for (let i = 0; i < args.length; i += 2) {
            if (args[i].startsWith('-')) {
                myArgs.set(args[i].substring(1), args[i + 1]);
            }
        }
        return myArgs;
    }
}

// Run the demo
const demo = new GoogleSheetsDemo();
demo.main(process.argv.slice(2));

function prompt(promptName, label, punctuation, defaultVal)
{
  lastPrompt = promptName;
  lastDefault = defaultVal;
  process.stdout.write(`${label} [${defaultVal}]${punctuation} `);
}

File: Cloud Sheets Samples/Google Sheets/c++/googlesheets.cpp
/*
 * Cloud Sheets 2024 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Sheets in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudsheets
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */


#include <iostream>
#include "../../include/cloudsheets.h"
#include <unordered_map>
#include <string>
#include <cstring>

#define BUFF_SIZE 500
bool changesMade = false;

class MyGoogleSheets : public GoogleSheets
{
public:
	virtual int FireSpreadsheetList(GoogleSheetsSpreadsheetListEventParams* e) {
		printf("Id: %s\n", e->Id);
		printf("Name: %s\n\n", e->Name);

		return 0;
	}
};
std::unordered_map<std::string, std::string> parseArgs(int argc, char* argv[]);
char* prompt(char* prompt);
void printMenu();
void toLowerCase(char* str);
void authenticate(const char* client, const char* secret);

const char* oAuthScopes = "https://www.googleapis.com/auth/spreadsheets https://www.googleapis.com/auth/drive";
const char* oAuthAuthUrl = "https://accounts.google.com/o/oauth2/auth";
const char* oAuthTokenUrl = "https://accounts.google.com/o/oauth2/token";

MyGoogleSheets gsheets;

int main(int argc, char* argv[]) {
	if (argc < 4)
	{
		printf("usage: GoogleSheetsDemo.cpp -c client -s secret\n" \
			" -c The Id of the client assigned when registering the application.\n" \
			" -s The secret value for the client assigned when registering the application.");

		return 0;
	}

	std::unordered_map<std::string, std::string> myArgs = parseArgs(argc, argv);
	char buffer[BUFF_SIZE];

	authenticate(myArgs.at("c").c_str(), myArgs.at("s").c_str());

	printf("Listing spreadsheets...\n");
	gsheets.ListSpreadsheets();

	while (true)
	{
		int retCode = 0;
		printf("\nEnter command: ");
		scanf("%[^\n]", buffer);

		if (!strcmp(buffer, "ls"))
		{
			retCode = gsheets.ListSpreadsheets();
			if (retCode != 0)
			{
				printf("Error: [%i] %s\n\n", retCode, gsheets.GetLastError());
			}
			else
			{
				printf("Success...");
			}
		}
		else if (!strcmp(buffer, "lssh"))
		{
			char spreadsheetId[BUFF_SIZE];
			printf("Enter the Spreadsheet Id: ");
			getchar();
			scanf("%[^\n]", spreadsheetId);

			retCode = gsheets.LoadSpreadsheet(spreadsheetId);
			if (retCode != 0)
			{
				printf("Error: [%i] %s\n\n", retCode, gsheets.GetLastError());
			}
			else
			{
				for (int i = 0; i < gsheets.GetSheets()->GetCount(); i++)
				{
					printf("Id: %d\n", gsheets.GetSheets()->Get(i)->GetSheetId());
					printf("Title: %s\n", gsheets.GetSheets()->Get(i)->GetTitle());
					printf("Index: %d\n", gsheets.GetSheets()->Get(i)->GetSheetIndex());
					printf("Total Columns: %d\n", gsheets.GetSheets()->Get(i)->GetTotalColumnCount());
					printf("Total Rows: %d\n\n", gsheets.GetSheets()->Get(i)->GetTotalRowCount());
				}
			}
		}
		else if (!strcmp(buffer, "cs"))
		{
			char name[BUFF_SIZE], locale[BUFF_SIZE], timezone[BUFF_SIZE];

			printf("Enter spreadsheet name: ");
			getchar();
			scanf("%[^\n]", name);

			printf("Enter locale (e.g. en_US): ");
			getchar();
			scanf("%[^\n]", locale);

			printf("Enter timezone (e.g. America/New_York): ");
			getchar();
			scanf("%[^\n]", timezone);
			retCode = gsheets.CreateSpreadsheet(name, locale, timezone);

			if (retCode != 0)
			{
				printf("Error: [%i] %s\n\n", retCode, gsheets.GetLastError());
			}
			else
			{
				printf("Success...");
			}
		}
		else if (!strcmp(buffer, "ash"))
		{
			char spreadsheetId[BUFF_SIZE];
			printf("Enter the Spreadsheet Id: ");
			getchar();
			scanf("%[^\n]", spreadsheetId);

			retCode = gsheets.LoadSpreadsheet(spreadsheetId);
			if (retCode != 0)
			{
				printf("Error: [%i] %s\n\n", retCode, gsheets.GetLastError());
			}
			else 
			{
				char sheetIndexStr[BUFF_SIZE];
				printf("Enter the index where sheet should be added: ");
				getchar();
				scanf("%[^\n]", sheetIndexStr);
				int sheetIndex = atoi(sheetIndexStr);

				char sheetTitle[BUFF_SIZE];
				printf("Enter sheet title: ");
				getchar();
				scanf("%[^\n]", sheetTitle);

				retCode = gsheets.AddSheet(sheetIndex, sheetTitle);
				if (retCode != 0)
				{
					printf("Error: [%i] %s\n\n", retCode, gsheets.GetLastError());
				}
				else
				{
					printf("Success...");
				}
			}

		}
		else if (!strcmp(buffer, "us"))
		{
			changesMade = false;
			char spreadsheetId[BUFF_SIZE];
			printf("Enter the Spreadsheet Id: ");
			getchar();
			scanf("%[^\n]", spreadsheetId);

			retCode = gsheets.LoadSpreadsheet(spreadsheetId);
			if (retCode != 0)
			{
				printf("Error: [%i] %s\n\n", retCode, gsheets.GetLastError());
			}
			else
			{
				CloudSpreadsheetsGSSpreadsheet* spreadsheet = gsheets.GetSpreadsheets()->Get(0);

				printf("Current name: %s\n", spreadsheet->GetName());
				if (strcmp(prompt("Change? (y/n): "), "y") == 0)
				{
					const char* newName = prompt("Enter new name: ");

					gsheets.GetSpreadsheets()->Get(0)->SetName(newName);
					changesMade = true;
				}

				printf("Current locale: %s\n", spreadsheet->GetLocale());
				if (strcmp(prompt("Change? (y/n): "), "y") == 0)
				{
					const char* newLocale = prompt("Enter new locale (e.g. en_US): ");
					gsheets.GetSpreadsheets()->Get(0)->SetLocale(newLocale);
					changesMade = true;
				}

				printf("Current timezone: %s\n", spreadsheet->GetTimeZone());
				if (strcmp(prompt("Change? (y/n): "), "y") == 0)
				{
					const char* newTimeZone = prompt("Enter new timezone (e.g. America/New_York): ");
					gsheets.GetSpreadsheets()->Get(0)->SetTimeZone(newTimeZone);
					changesMade = true;
				}

				if (changesMade)
				{
					retCode = gsheets.UpdateSpreadsheet();
					if (retCode != 0)
					{
						printf("Error: [%i] %s\n\n", retCode, gsheets.GetLastError());
					}
					else
					{
						printf("Success... Spreadsheet updated!\n");
					}
					changesMade = false;
				}
				else
				{
					printf("No changes were made to the spreadsheet.\n");
				}
			}
		}
		else if (!strcmp(buffer, "ush"))
		{
			changesMade = false;
			char currentSpreadsheetId[BUFF_SIZE];
			printf("Enter Spreadsheet Id: ");
			getchar();
			scanf("%[^\n]", currentSpreadsheetId);

			char currentSheetIdStr[BUFF_SIZE];
			printf("Enter Sheet Id (Should be in Spreadsheet with Id: %s): ", currentSpreadsheetId);
			getchar();
			scanf("%[^\n]", currentSheetIdStr);
			int currentSheetId = atoi(currentSheetIdStr);

			retCode = gsheets.LoadSpreadsheet(currentSpreadsheetId);
			if (retCode != 0)
			{
				printf("Error: [%i] %s\n\n", retCode, gsheets.GetLastError());
			}

			retCode = gsheets.LoadSheet(currentSheetId);
			
			if (retCode != 0)
			{
				printf("Error: [%i] %s\n\n", retCode, gsheets.GetLastError());
			}
			else 
			{
				CloudSpreadsheetsGSSheet* sheet = gsheets.GetSheets()->Get(0);

				printf("Current name: %s\n", sheet->GetTitle());
				if (strcmp(prompt("Change? (y/n): "), "y") == 0)
				{
					const char* newTitle = prompt("Enter new title: ");
					gsheets.GetSheets()->Get(0)->SetTitle(newTitle);
					changesMade = true;
				}

				printf("Current index: %d\n", sheet->GetSheetIndex());
				if (strcmp(prompt("Change? (y/n): "), "y") == 0)
				{
					const char* newIndex_input = prompt("Enter new index: ");
					int newIndex = atoi(newIndex_input);
					gsheets.GetSheets()->Get(0)->SetSheetIndex(newIndex);
					changesMade = true;
				}

				if (changesMade)
				{
					retCode = gsheets.UpdateSheet();
					if (retCode != 0)
					{
						printf("Error: [%i] %s\n\n", retCode, gsheets.GetLastError());
					}
					else
					{
						printf("Success... Sheet updated!\n");
					}
					changesMade = false;
				}
				else
				{
					printf("No changes were made to the sheet.\n");
				}
			}
		}
		else if (!strcmp(buffer, "ds"))
		{
			char spreadsheetId[BUFF_SIZE];
			printf("Enter the Spreadsheet Id: ");
			getchar();
			scanf("%[^\n]", spreadsheetId);

			retCode = gsheets.DeleteSpreadsheet(spreadsheetId);
			if (retCode != 0)
			{
				printf("Error: [%i] %s\n\n", retCode, gsheets.GetLastError());
			}
			else
			{
				printf("Success...");
			}
		}
		else if (!strcmp(buffer, "dsh"))
		{
			char spreadsheetId[BUFF_SIZE];
			printf("Enter the Spreadsheet Id: ");
			getchar();
			scanf("%[^\n]", spreadsheetId);

			retCode = gsheets.LoadSpreadsheet(spreadsheetId);
			if (retCode != 0)
			{
				printf("Error: [%i] %s\n\n", retCode, gsheets.GetLastError());
			}

			char sheetId_input[BUFF_SIZE];
			printf("Enter the Sheet Id: ");
			getchar();
			scanf("%[^\n]", sheetId_input);
			int sheetId = atoi(sheetId_input);

			gsheets.DeleteSheet(sheetId);
			if (retCode != 0)
			{
				printf("Error: [%i] %s\n\n", retCode, gsheets.GetLastError());
			}
			else
			{
				printf("Success...");
			}

		}
		else if (!strcmp(buffer, "exp"))
		{
			printf("Supported file extensions: csv, ods, pdf, tsv, xlsx, zip\n");

			char currentSpreadsheetId[BUFF_SIZE];
			printf("Enter Spreadsheet Id: ");
			getchar();
			scanf("%[^\n]", currentSpreadsheetId);

			char fileName[BUFF_SIZE];
			printf("Enter full filename (e.g. myFile.zip): ");
			getchar();
			scanf("%[^\n]", fileName);

			char filePath[BUFF_SIZE];
			printf("Enter save directory: ");
			getchar();
			scanf("%[^\n]", filePath);

			snprintf(filePath, BUFF_SIZE, "%s\\%s", filePath, fileName);
			printf("File will be saved at: %s\n", filePath);

			retCode = gsheets.Export(currentSpreadsheetId, filePath);
			if (retCode != 0)
			{
				printf("Error: [%i] %s\n\n", retCode, gsheets.GetLastError());
			}
			else
			{
				printf("Success... File exported successfully.\n");
			}
		}
		else if (!strcmp(buffer, "q"))
		{
			return 0;
		}
		else if (!strcmp(buffer, "?"))
		{
			printMenu();
		}
		else
		{
			printf("Command not recognized!");
			printMenu();
		}
		getchar();
	}
}

void printMenu()
{
	printf("\r\n?\t-\tHelp\n" \
		"ls\t-\tList Spreadsheets\n" \
		"lssh\t-\tList Sheets\n" \
		"cs\t-\tCreate Spreadsheet\n" \
		"ash\t-\tAdd Sheet\n" \
		"us\t-\tUpdate Spreadsheet\n" \
		"ush\t-\tUpdate Sheet\n" \
		"ds\t-\tDelete Spreadsheet\n" \
		"dsh\t-\tDelete Sheet\n" \
		"exp\t-\tExport Spreadsheet\n" \
		"q\t-\tQuit\n");
}

void authenticate(const char* client, const char* secret)
{
	OAuth oauth;

	oauth.SetClientId(client);
	oauth.SetClientSecret(secret);
	oauth.SetAuthorizationScope(oAuthScopes);
	oauth.SetServerAuthURL(oAuthAuthUrl);
	oauth.SetServerTokenURL(oAuthTokenUrl);

	gsheets.SetAuthorization(oauth.GetAuthorization());
}

void toLowerCase(char* str)
{
	for (size_t i = 0; i < std::strlen(str); ++i) {
		str[i] = std::tolower(static_cast<unsigned char>(str[i]));
	}
}

char* prompt(char* prompt)
{
	static char buffer[BUFF_SIZE];

	printf("%s", prompt);
	getchar();
	scanf("%[^\n]", buffer);

	return buffer;
}


std::unordered_map<std::string, std::string> parseArgs(int argc, char* argv[])
{
	std::unordered_map<std::string, std::string> map;
	for (int i = 0; i < argc; i++)
	{
		if (argv[i][0] == '-')
		{
			if (i + 1 < argc && (argv[i + 1][0] != '-'))
			{
				char* newStr = argv[i] + 1;
				toLowerCase(newStr);
				map[newStr] = argv[i + 1];
				i++;
			}
			else
			{
				char* newStr = argv[i] + 1;
				toLowerCase(newStr);
				map[newStr] = "";
			}
		}
		else
		{
			map[std::to_string(i)] = argv[i];
		}
	}
	return map;
}

